packages/@react-spectrum/button/src/Button.tsx

```diff
@@ -126,22 +139,34 @@ function Button<T extends ElementType = 'button'>(props: SpectrumButtonProps<T>,
         <SlotProvider
           slots={{
             icon: {
+              id: iconId,
               size: 'S',
               UNSAFE_className: classNames(styles, 'spectrum-Icon')
             },
             text: {
+              id: textId,
               UNSAFE_className: classNames(styles, 'spectrum-Button-label')
             }
           }}>
-          {isProgressVisible && <ProgressCircle
-            aria-label={stringFormatter.format('loading')}
-            isIndeterminate
-            size="S"
-            UNSAFE_className={classNames(styles, 'spectrum-Button-circleLoader')}
-            staticColor={staticColor} />}
           {typeof children === 'string'
             ? <Text>{children}</Text>
             : children}
+          {isPending && <ProgressCircle
+            aria-hidden="true"
+            isIndeterminate
+            size="S"
+            UNSAFE_className={classNames(styles, 'spectrum-Button-circleLoader')}
+            UNSAFE_style={{visibility: isProgressVisible ? 'visible' : 'hidden'}}
+            staticColor={staticColor} />
+          }
+          {isPending && hasAriaLabel &&
+            <div aria-hidden="true" id={auxLabelId} aria-label={buttonProps['aria-label']} aria-labelledby={buttonProps['aria-labelledby']?.replace(buttonId, auxLabelId)} />
+          }
+          {isPending && <div
+            id={spinnerId}
+            aria-label={stringFormatter.format('pending')}
+            aria-labelledby={`${hasAriaLabel ? '' : iconId} ${hasAriaLabel ? auxLabelId : textId} ${spinnerId}`} />
+          }
```

Comment:
Some code comments about why all this complexity exists and what screen reader/browser issues its working around would be good.